### PR TITLE
Jim Riddle

### PR DESCRIPTION
**Pull request by jimriddle1**

### CODE REVIEW

File: lib/museum.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        #do nothing
+      else
+        patron.charge_patron(exhibit.cost)
+        @patrons_of_exhibits[exhibit] << patron
+        @revenue += exhibit.cost
+      end
+    end
+  end
+
+  def recommended_exhibits(patron)
```
Comment (cjsim89) [inline lib/museum.rb:43]: Careful of this name - the pattern called it `recommend_exhibits`, which is a very small detail but those details can add up.

File: lib/museum.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def add_exhibit(exhibit)
+    @exhibits << exhibit
+  end
+
+  def admit(patron)
+    @patrons << patron
+    affordable_exhibits = []
+    recommended_exhibits(patron).each do |exhibit|
```
Comment (cjsim89) [inline lib/museum.rb:20]: So, the interaction pattern has the name of that method as `recommend_exhibits`, and it wouldn't be a big deal but since you're calling it from multiple places it can be a headache to correct. Just keep in mind the details that need to be used from the pattern - copying & pasting is a good idea sometimes.

File: lib/museum.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      affordable_exhibits << exhibit
+    end
+
+    ordered_exhibits = affordable_exhibits.sort_by{ |exhibit| exhibit.cost}.reverse
```
Comment (cjsim89) [inline lib/museum.rb:24]: Great use of enumerables and built-in methods!

File: lib/museum.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    end
+
+    ordered_exhibits.each do |exhibit|
+      if exhibit.cost > patron.spending_money
+        #do nothing
+      else
+        patron.charge_patron(exhibit.cost)
+        @patrons_of_exhibits[exhibit] << patron
+        @revenue += exhibit.cost
+      end
```
Comment (cjsim89) [inline lib/museum.rb:39]: I respect the `if/else` here; but is there a better way to write the condition so that only the `else` codeblock is executed?

File: lib/museum.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    end
+  end
+
+  def recommended_exhibits(patron)
+    recommended = []
+    exhibits.each do |exhibit|
+      patron.interests.each do |interest|
+        # require 'pry'; binding.pry
```
Comment (cjsim89) [inline lib/museum.rb:47]: Nice job using `each` to start with - try to come back and figure out a way to refactor this to use other enumerables, for good practice!

File: lib/museum.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        if interest == exhibit.name
+          recommended << exhibit
+        end
+      end
+    end
+    return recommended
+  end
+
+  def patrons_by_exhibit_interest
+    hash = {}
```
Comment (cjsim89) [inline lib/museum.rb:57]: This syntax to make a new hash is fine, but it could cause problems when you're accessing values because this will make new values as `nil`. Try `Hash.new(0)` instead, which sets empty values to `0` which can be easier to access, causes fewer weird errors when debugging.

File: lib/museum.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    @exhibits.each do |exhibit|
+      hash[exhibit] = []
+    end
+    @exhibits.each do |exhibit|
+      @patrons.each do |patron|
+        if patron.interests.include?(exhibit.name)
+          hash[exhibit] << patron
+        end
+      end
+    end
```
Comment (cjsim89) [inline lib/museum.rb:67]: Again, great job with nested iteration here. Try to come back and refactor using other enumerables to get good practice.

File: spec/museum_spec.rb [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+    dmns.admit(patron_1)
+    dmns.admit(patron_2)
+    dmns.admit(patron_3)
+
+    expected = {}
+
+    expected[gems_and_minerals] = [patron_1]
+    expected[dead_sea_scrolls] = [patron_1, patron_2, patron_3]
+    expected[imax] = []
```
Comment (cjsim89) [inline spec/museum_spec.rb:111]: Nice job setting up the hash this way! Makes the actual test statement easier to read.

File: spec/museum_spec.rb [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+    patron_3 = Patron.new("Johnny", 5)
+    patron_3.add_interest("Dead Sea Scrolls")
+
+    dmns.admit(patron_1)
+    dmns.admit(patron_2)
+    dmns.admit(patron_3)
+
+    # expect(dmns.draw_lottery_winner(dead_sea_scrolls)).to eq("Johnny" || "Bob")
+    expect(dmns.draw_lottery_winner(gems_and_minerals)).to eq(nil)
```
Comment (cjsim89) [inline spec/museum_spec.rb:172]: This is fine, but missing a test to see when the lottery winner is _not_ nil.

File: spec/museum_spec.rb [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    patron_2.add_interest("IMAX")
+    patron_2.add_interest("Dead Sea Scrolls")
+    dmns.admit(patron_2)
+
+    morgan = Patron.new("Morgan", 15)
+    morgan.add_interest("Gems and Minerals")
+    morgan.add_interest("Dead Sea Scrolls")
+    dmns.admit(morgan)
+
+    expect(dmns.revenue).to eq(35)
```
Comment (cjsim89) [inline spec/museum_spec.rb:290]: Nice job with these i4 tests! These were a doozy, eh?

**Review by cjsim89 — COMMENTED** [review]
Jim - really great work with this week's IC. I left some comments below. Keep up the good work with testing, committing, and enumerables!
